src/validators/UnionValidator.ts

```diff
@@ -56,19 +56,24 @@ export class UnionValidator<T> extends BaseValidator<T> {
 			return this.clone();
 		}
 
-		return new UnionValidator([new LiteralValidator(null), this.clone()]);
+		return new UnionValidator([new LiteralValidator(null), ...this.validators]);
 	}
 
 	public override get nullish(): UnionValidator<T | null | undefined> {
 		if (this.validators.length === 0) return new UnionValidator<T | null | undefined>([new NullishValidator()], this.constraints);
 
 		const [validator] = this.validators;
-		if (validator instanceof NullishValidator) {
+		if (validator instanceof LiteralValidator) {
+			// If already nullable or optional, promote the union to nullish:
+			if (validator.expected === null || validator.expected === undefined) {
+				return new UnionValidator<T | null | undefined>([new NullishValidator(), ...this.validators.slice(1)], this.constraints);
+			}
```

Comment:
Handle only undefined and only null too? Because this will changes the behavior of s.union(s.literal(null), s.string) from string | null to string | undefined | null